File: test/engine.py
"""Neovim abstraction for tests."""

import logging
import os
import threading
import time
from pynvim import attach
from spy_ui import SpyUI


# Neovim proxy
class Engine:
    """Neovim engine."""

    def __init__(self):
        """Construct an engine."""
        self.logger = logging.getLogger("Engine")
        self.logger.setLevel(logging.DEBUG)
        lhandl = logging.NullHandler() if not os.environ.get('CI') \
            else logging.FileHandler("engine.log")
        fmt = "%(asctime)s [%(levelname)s]: %(message)s"
        lhandl.setFormatter(logging.Formatter(fmt))
        self.logger.addHandler(lhandl)

        self.logger.info("Starting test nvim engine")

        self.screen = ""

        args = ["/usr/bin/env", "./nvim", "--embed", "--headless", "-n",
                "--listen", "localhost:44444"]
        self.nvim = attach('child', argv=args)
        self.spy_ui = None
        self.thrd = threading.Thread(target=self.run_ui)
        self.thrd.start()
        # Dummy request to make sure the embedded Nvim proceeds
        # See in neovim bd8d43c6fef868 (startup: wait for embedder
        # before executing)
        self.eval("0")

        # Builds on GitHub seem to be more prone to races.
        is_github = os.environ.get('GITHUB_WORKFLOW')
        self.feed_delay = 0.02 if not is_github else 0.5
        self.launch_delay = 3000 if not is_github else 20000

    def close(self):
        '''Close.'''
        self.spy_ui.close()
        try:
            self.nvim.command(":qa!")
        except OSError:
            pass
        self.nvim.close()
        self.thrd.join()

    def log_screen(self):
        """Log the current Spy UI screen if it has changed."""
        if not self.spy_ui:
            return
        screen = self.spy_ui.screen
        if screen != self.screen:
            self.screen = screen
            self.logger.info("\n%s", screen)

    def run_ui(self):
        """Capture neovim UI."""
        self.spy_ui = SpyUI()
        self.spy_ui.run()

    def exe(self, cmd, delay=100):
        """Execute a Vim command."""
        self.log_screen()
        self.logger.info("exe «%s»", self._quote_keys(cmd))
        self.nvim.command(cmd)
        time.sleep(delay * 0.001)
        self.log_screen()

    def get_signs(self):
        """Get pointer position and list of breakpoints."""

        ret = {}

        for buf in self.nvim.buffers:
            if not buf.valid or not self.nvim.api.buf_is_loaded(buf.handle):
                continue
            out = self.eval(f'sign_getplaced({buf.number}, {{"group": "NvimGdb"}})')
            breaks = {}
            for bsigns in out:
                for signs in bsigns['signs']:
                    sname = signs['name']
                    if sname == 'GdbCurrentLine':
                        bname = os.path.basename(buf.name)
                        assert "cur" not in ret
                        ret["cur"] = f'{bname}:{signs["lnum"]}'
                    if sname.startswith('GdbBreakpoint'):
                        num = int(sname[len('GdbBreakpoint'):])
                        try:
                            breaks[num].append(signs["lnum"])
                        except KeyError:
                            breaks[num] = [signs["lnum"]]
            if breaks:
                ret['break'] = breaks
        return ret

    def feed(self, keys, delay=100):
        """Send a Vim keystroke to NeoVim."""
        time.sleep(self.feed_delay)
        self.log_screen()
        self.logger.info("feed «%s»", self._quote_keys(keys))
        self.nvim.input(keys)
        time.sleep(delay * 0.001)
        self.log_screen()

    def eval(self, expr):
        """Evaluate a Vim expression."""
        self.log_screen()
        self.logger.info("eval «%s»", expr)
        return self.nvim.eval(expr)

    def exec_lua(self, expr):
        """Execute lua statement."""
        self.log_screen()
        self.logger.info("exec_lua «%s»", expr)
        return self.nvim.exec_lua(expr)

    def count_buffers(self):
        """Determine how many buffers are there."""
        self.eval('len(filter(nvim_list_bufs(), "nvim_buf_is_loaded(v:val)"))')

    def count_termbuffers(self):
        """Determine how many terminal buffers are there."""
        terms = [b for b in self.nvim.buffers
                 if b.api.is_loaded() and
                 b.api.get_option('buftype') == 'terminal']
        return len(terms)

    @staticmethod
    def wait_for(action, condition, deadline=2000):
        '''Wait until the action returns the expected value.'''
        deadline *= 0.001
        start = time.time()
        result = None
        while time.time() - start <= deadline:
            result = action()
            if condition(result):
                return None
            time.sleep(0.2)
        return result

    def wait_signs(self, expected, deadline=2000):
        '''Wait until signs are placed as expected.'''
        return self.wait_for(self.get_signs,
                             lambda res: res == expected, deadline)

    def wait_paused(self):
        '''Wait until the parser FSM goes into the paused state.'''
        return self.wait_for(
            lambda: self.exec_lua("return nvimgdb.i().parser:is_paused()"),
            lambda res: res, self.launch_delay)

    @staticmethod
    def _quote_keys(keys):
        return keys.replace('\n', '\\n').replace('\r', '\\r') \
            .replace('\t', '\\t').replace('\b', '\\b')

File: test/engine.lua
local thr = require'thread'
local E = {}

---Feed keys to Neovim
---@param keys string @keystrokes
---@param timeout? number @delay in milliseconds after the input
function E.feed(keys, timeout)
  vim.api.nvim_input(keys)
  thr.y(timeout == nil and 200 or timeout)
end

function E.exe(cmd)
  thr.y(0, vim.cmd(cmd))
end

function E.get_time_ms()
  return vim.loop.hrtime() * 1e-6
end

function E.wait_for(query, check, timeout_ms)
  if timeout_ms == nil then
    timeout_ms = 5000
  end
  local deadline = E.get_time_ms() + timeout_ms
  local value = nil
  while E.get_time_ms() < deadline do
    value = query()
    if check(value) then
      return true
    end
    thr.y(100)
  end
  return value
end

---Wait until the debugger gets into the paused state
---@param timeout_ms? number Timeout in milliseconds
---@return boolean
function E.wait_paused(timeout_ms)
  if timeout_ms == nil then
    timeout_ms = 5000
  end
  local query = function()
    if NvimGdb == nil then
      return false
    end
    local parser = NvimGdb.i().parser
    return type(parser) == 'table' and parser:is_paused()
  end
  local function is_true(v)
    return v
  end
  return E.wait_for(query, is_true, timeout_ms)
end

function E.count_buffers_impl(pred)
  local count = 0
  for _, buf in ipairs(vim.api.nvim_list_bufs()) do
    if pred(buf) then
      count = count + 1
    end
  end
  return count
end

function E.count_buffers()
  -- Determine how many terminal buffers are there.
  return E.count_buffers_impl(function(buf)
    return vim.api.nvim_buf_is_loaded(buf)
  end)
end

function E.count_termbuffers()
  -- Determine how many terminal buffers are there.
  return E.count_buffers_impl(function(buf)
    return vim.api.nvim_buf_is_loaded(buf) and vim.api.nvim_buf_get_option(buf, 'buftype') == 'terminal'
  end)
end

---Get current signs: current line and breakpoints
---@return table
function E.get_signs()
  -- Get pointer position and list of breakpoints.
  local ret = {}

  for _, buf in ipairs(vim.api.nvim_list_bufs()) do
    if vim.api.nvim_buf_is_valid(buf) and vim.api.nvim_buf_is_loaded(buf) then
      local breaks = {}
      for _, bsigns in ipairs(vim.fn.sign_getplaced(buf, {group = "NvimGdb"})) do
        for _, signs in ipairs(bsigns.signs) do
          local sname = signs.name
          if sname == 'GdbCurrentLine' then
            local bname = vim.api.nvim_buf_get_name(buf):match("[^/\\]+$")
            if ret.cur == nil then
              ret.cur = bname .. ':' .. signs.lnum
            else
              if ret.curs == nil then
                ret.curs = {}
              end
              table.insert(ret.curs, bname .. ':' .. signs.lnum)
            end
          end
          if sname:match('^GdbBreakpoint') then
            local num = assert(tonumber(sname:sub(1 + string.len('GdbBreakpoint'))))
            if breaks[num] == nil then
              breaks[num] = {}
            end
            table.insert(breaks[num], signs.lnum)
          end
        end
      end
      if next(breaks) ~= nil then
        ret.brk = breaks
      end
    end
  end
  return ret
end

function E.wait_signs(expected_signs, timeout_ms)
  local function query()
    return E.get_signs()
  end
  local function is_expected(signs)
    return vim.deep_equal(expected_signs, signs)
  end
  return E.wait_for(query, is_expected, timeout_ms)
end

return E
